File: 142. Linked List Cycle II/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # edge condition
        if not head or not head.next:
            return None
        
        # detect cycle
        fptr = head
        sptr = head
        cycle = False
        while fptr and fptr.next and not cycle:
            fptr = fptr.next.next
            sptr = sptr.next
            cycle = fptr == sptr
            
        if not cycle:
            return None
        
        # detect start
        fptr = head
        while fptr != sptr:
            fptr = fptr.next
            sptr = sptr.next
            
        return sptr
        
            
        

File: 142. Linked List Cycle II/solution.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null || head.next == null) {
            return null;
        }
        
        ListNode sptr = head;
        ListNode fptr = head;
        
        while (fptr != null && fptr.next != null) {
            fptr = fptr.next.next;
            sptr = sptr.next;
            if (fptr == sptr) {
                break;
            }
        }

        if (fptr == null || fptr.next == null) {
            return null;
        }

        sptr = head;
        while (sptr != fptr) {
            fptr = fptr.next;
            sptr = sptr.next;
        }
        
        return sptr;
    }
}
